### PR TITLE
CP-24919: Specify a timeout for IO with client.

### PR DESCRIPTION
**Pull request by thomassa**
This will get rid of idle clients.

### CODE REVIEW

**Comment by thomassa** [comment]
This pullreq depends on pullreq xapi-project/nbd#114

**Comment by thomassa** [comment]
I've tested this: it does indeed close an idle connection after the specified time, and activity on the connection resets the timer.

File: src/main.ml
Comment (gaborigloi) [inline src/main.ml]: This can be simplified to:
```ocaml
inc_conn () >>= fun () ->
Lwt_unix.setsockopt fd Lwt_unix.SO_KEEPALIVE true;
xapi_says_use_tls () >>= fun tls ->
let tls_role = if tls then tls_server_role else None in
handle_connection fd tls_role
```
Here we're using the idiomatic `Lwt` formatting: we don't indent the lines after the `bind`s (the `>>=`s).

Comment (thomassa) [reply src/main.ml]: No, because `xapi_says_use_tls` takes `unit`, so that code gives
```
File "src/main.ml", line 151, characters 22-39:
Error: This expression has type unit -> bool Lwt.t
       but an expression was expected of type 'a Lwt.t
```
referring to characters `xapi_says_use_tls`.

But we can fix this by changing the `;` in your code snippet to `|>` so I'll do that.

Comment (thomassa) [reply src/main.ml]: Or maybe it would be clearer to have
```
...;
xapi_says_use_tls () >>=
```
rather than
```
... |>
xapi_says_use_tls >>=
```

Comment (gaborigloi) [reply src/main.ml]: Oops, I've update my comment - I think it's much clearer if we explicitly pass the `()` to it.

File: src/main.ml
Comment (gaborigloi) [inline src/main.ml]: Could you please rewrite this block as well in the usual Lwt style? Thanks!

**Review by gaborigloi — APPROVED** [review]
Looks good. It's sad that the method using [`setsockopt SO_RCVTIMEO`](https://github.com/gaborigloi/xapi-nbd/tree/idle_timeout) did not work :( we should investigate why at some point.
First the `nbd` pull request should be merged.

**Comment by thomassa** [comment]
I've squashed as appropriate, so this should be ready to merge once the `nbd` pull-request has been merged.
(If we're tracking the 2.x branch of nbd, anyway... otherwise we'll also need a new release.)

File: lib/consts.ml
Comment (mseri) [inline lib/consts.ml:30]: Isn't 17s somewhat conservative?

Comment (thomassa) [reply lib/consts.ml:30]: @mseri ?

Comment (mseri) [reply lib/consts.ml:30]: Most of our timeouts are on the order of minutes, I was wandering why 17 seconds. I don't have strong opinions. If you are all happy with it, it's fine

**Comment by mseri** [comment]
Is there any reason this is not yet merged?

**Comment by gaborigloi** [comment]
It depends on https://github.com/xapi-project/nbd/pull/114 , for which I wanted to measure the performance impact first. Also, should we increase the timeout?

File: lib/consts.ml
Comment (gaborigloi) [reply lib/consts.ml:30]: @edwintorok @thomasmck @robertbreker What do you think about this timeout? Should it be longer?
I think it should be longer.

Comment (gaborigloi) [reply lib/consts.ml:30]: The ticket originally said that the timeout should be 5 minutes.

Comment (edwintorok) [reply lib/consts.ml:30]: FWIW nginx uses 1 minute, but how many bytes do you read/write in each of these requests? If it is X bytes / 17s you can calculate what speed that would be, and if it turns out a modem could meet that speed then no need to increase it...

**Comment by mseri** [comment]
We decided to drop the timeout there. Is there anything that we need to do here?

**Comment by gaborigloi** [comment]
We can temporarily close this and then repoen when things are ready on the nbd side?

**Comment by thomassa** [comment]
@mseri Why drop the timeout? (And do you mean remove or reduce?)

**Comment by mseri** [comment]
No, I think it was decided for the moment to not have a timeout because it was for the whole operation and was too risky. @gaborigloi can tell you more

**Comment by gaborigloi** [comment]
Closing for now as the original PR was closed. Will re-open once it gets into `nbd`.

**Pull request closed without merging**
